File: python/src/actr_harness/services/neuro_core/conditions/fuzzy_condition_evaluator.py
import logging

from actr_harness.services.llm_client import LLMClient
from actr_harness.observability import log_event, observe_boundary
from ..buffers_view import BuffersView
from actr_harness.generated.grpc.actr import (
    ProceduralCondition,
)

logger = logging.getLogger(__name__)


class FuzzyConditionEvaluator:
    def __init__(self, llm: LLMClient):
        self._llm = llm

    @observe_boundary("fuzzy_condition_evaluator.evaluate")
    async def evaluate(self, conditions: list[ProceduralCondition], view: BuffersView) -> list[str]:
        prompt = {
            "buffers": view.to_dict(),
            "conditions": [
                {
                    "rule_id": c.rule_id,
                    "symbolic": c.condition.to_dict(),
                    "semantic_hint": c.semantics.to_dict(),
                }
                for c in conditions
            ]
        }
        system = (
            "You are given buffers (world state) and conditions with optional semantic hints. "
            "Determine which conditions are satisfied. "
            "Return ONLY a JSON array of the satisfied rule_id strings. "
            "No extra text, no explanation."
        )
        result = await self._llm.chat_json(prompt, system)
        if not isinstance(result, list):
            log_event(
                logger,
                logging.WARNING,
                "fuzzy_condition_evaluation.invalid_payload",
                "Fuzzy condition evaluator returned a non-list payload.",
                payload_type=type(result).__name__,
            )
            return []

        log_event(
            logger,
            logging.DEBUG,
            "fuzzy_condition_evaluation.completed",
            "Resolved fuzzy conditions.",
            candidate_rule_ids=[condition.rule_id for condition in conditions],
            matched_rule_ids=result,
        )
        return result

File: Dotnet/src/Harness.Core/NeuroCore/FuzzyConditionEvaluator.cs
using System.Text.Json.Nodes;
using Harness.Abstractions.Actr;
using Harness.Shared.Observability;
using Microsoft.Extensions.Logging;

namespace Harness.Core.NeuroCore;

public class FuzzyConditionEvaluator(LlmClient llm, ILogger<FuzzyConditionEvaluator> logger) : IProvideLogger
{
    public ILogger Logger => logger;

    [ObserveBoundary]
    public async Task<IReadOnlyList<string>> EvaluateAsync(
        IReadOnlyList<ProceduralCondition> conditions,
        BuffersView view,
        CancellationToken cancellationToken = default)
    {
        var prompt = new Dictionary<string, object?>
        {
            ["buffers"] = view.ToDictionary(),
            ["conditions"] = conditions.Select(condition => new Dictionary<string, object?>
            {
                ["rule_id"] = condition.RuleId,
                ["symbolic"] = ProtobufDataConverter.ToPlainObjectMap(condition.Condition),
                ["semantic_hint"] = ProtobufDataConverter.ToPlainObjectMap(condition.Semantics)
            }).ToList()
        };

        const string systemPrompt =
            "You are given buffers (world state) and conditions with optional semantic hints. "
            + "Determine which conditions are satisfied. "
            + "Return ONLY a JSON array of the satisfied rule_id strings. "
            + "No extra text, no explanation.";

        var result = await llm.ChatJsonAsync(prompt, systemPrompt, cancellationToken);
        if (result is not JsonArray array)
        {
            logger.LogWarning("Fuzzy condition evaluator returned a non-list payload.");
            return [];
        }

        return array
            .Select(node => node?.GetValue<string>())
            .Where(ruleId => !string.IsNullOrWhiteSpace(ruleId))
            .Cast<string>()
            .ToList();
    }
}
